[2018.3] Strip trailing commas on Linux user's GECOS fields

meaksh opened the pull request:
### What does this PR do?
This PR strips the trailing commas (if any) created when the user GECOS fields (comma-delimited list) are generated by the `useradd` module on Linux systems.

Given this example state:
```yaml
create_my_test_user:
  user.present:
    - name: testuser
    - fullname: Testing User
```
After applying the state, the generated entry on `/etc/passwd` looks like this:
```
testuser:x:1002:100:Testing User,,,:/home/testuser:/bin/bash
```
Most of the times, the users are defined without `roomnumber`, `homephone` and `workphone` attributes so entries created on `/etc/passwd` are always containing those trailing commas.

On this cases the usual representation should be like:
```
mysql:x:60:466:MySQL database admin:/var/lib/mysql:/bin/false
```

So, after applying this PR, the entry generated on `/etc/passwd` would look like:
```
testuser:x:1002:100:Testing User:/home/testuser:/bin/bash
```

### Tests written?

Yes

### Commits signed with GPG?

Yes

meaksh committed:
Strip trailing commas on Linux user GECOS fields

meaksh committed:
Add unit tests for GECOS fields

terminalmage reviewed (commented):
Unless this is fixing a known bug, this change should not be made in the middle of a release cycle.

meaksh commented:
@terminalmage - This is actually fixing an issue reported by one of our customers. We already have this changes included in our Salt `2018.3.0` and `2016.11.4` packages.

Please let me know how to proceed here. Thanks! :+1:

terminalmage commented:
@cachedout should this be treated as a bug and included in 2018.3.1? My concern is that people who have `user.present` states which are currently returning `True` without making changes, will change the /etc/passwd once upgraded to 2018.3.1. We usually reserve these sort of changes for feature releases, when the behavior change isn't being made to resolve a bug affecting functionality. This seems more like a cosmetic change to me, though.

meaksh commented:
@terminalmage - AFAIK this change wouldn't produce any change on a `user.present` state that is currently returning `True`. No changes would be made on `/etc/passwd` after upgrading and applying `user.present` states because this module read first the GECOS fields from `/etc/passwd` and stores the values on a dict `{"workphone":.. "homephone":...}` which is the one used for making the comparison with the fields provided on the SLS file. Therefore, since there are no changes on "workphone", "homephone" etc, no changes will be detected and the state will keep returning `True`.

I agree with you that this is more like a cosmetic change than a bug itself.

cachedout commented:
I'm just not sure I can tell what the *bug* is here though. If this is cosmetic, as it seems to be, then I agree that this should not go into a point release. If it's a functional problem then we can continue this discussion?

isbm commented:
Seems like Salt is treating GECOS with commas only. @meaksh isn't that `chfn` can make a difference?

meaksh commented:
Oh, wait! I just noticed that `user.info` is not actually handling properly the GECOS fields on Linux (even without this PR):

1. salt-call user.add testuser
2. salt-call user.info testuser

Now we call `chfn` to add "other" attribute to GECOS fields for user `testuser`:

3. chfn -o foobar testuser

Then the `user.info` output is broken (notice the "homephone" attribute):

4. salt-call user.info testuser
```
local:
    ----------
    fullname:
    gid:
        100
    groups:
        - users
    home:
        /home/testuser
    homephone:
        ,foobar
    name:
        testuser
    passwd:
        x
    roomnumber:
    shell:
        /bin/bash
    uid:
        1001
    workphone:
```

I'll do the necessary changes here and will ping you back for a review. :+1:

isbm reviewed (changes requested):
@meaksh then I would rather fix all that. GECOS aren't exactly defined, and seems like trailing commas is OK to strip. However, this should be consistent with the external tools...

meaksh committed:
Fix unsupported chars checking on GECOS fields

meaksh committed:
Add 'other' attribute to GECOS fields to avoid inconsistencies with chfn

meaksh committed:
Add unit test for new method 'user.chother'

meaksh commented:
I pushed new commits to this PR to enable handling the `other` attribute on the GECOS fields. The previous behavior of Salt was adding all extra GECOS fields as part of the `homephone` field, but this is confusing and also inconsistent with `chfn` as pointed on https://github.com/saltstack/salt/pull/47149#issuecomment-383905343

Now, all extra GECOS fields are handled as an independent `other` attribute and `user.chother` function is now added to handle it. This `other` attribute will include all extra GECOS fields added after the fixed `fullname`, `roomnumber`, `workphone` and `homephone` fields.

Btw, I added https://github.com/saltstack/salt/pull/47149/commits/795b380ddb676c7ea32f9a5752da69a18f1867ae which actually fixes a bug which prevents checking if the GECOS fields contains unsupported characters.

@isbm @terminalmage @cachedout @rallytime - I think this is now ready for another review. Thanks!

isbm commented on salt/modules/useradd.py:
```diff
@@ -60,28 +60,30 @@ def _get_gecos(name):
+                                        gecos_dict.get('roomnumber', ''),
+                                        gecos_dict.get('workphone', ''),
+                                        gecos_dict.get('homephone', ''),
+                                        gecos_dict.get('other', ''),).rstrip(',')
```
I'd go here more pythonic way, actually:
```python
out = []
for field in ['fullname', 'roomnumber', 'workphone', 'homephone', 'other']:
    field = gecos_dict.get(field)
    if field:
        out.append(field)
return ','.join(out)
```

meaksh replied:
I would prefer to keep the current logic which is simpler and explicit rather than replacing it with a for loop + more variables and some extra calls, which unnecessarily makes the logic more complex.

isbm commented:
Arrgghh, I forgot to press "Finish review". :sob:

meaksh commented on salt/states/user.py:
```diff
@@ -448,6 +459,8 @@ def present(name,
         workphone = sdecode(workphone)
     if homephone is not None:
         homephone = sdecode(homephone)
+    if other is not None:
```
Yes, this behaving like the other GECOS attributes, in case of an empty string, it would be then stripped while rendering the GECOS fields.

meaksh commented on salt/states/user.py:
```diff
@@ -188,6 +189,12 @@ def _changes(name,
         if lusr['homephone'] != homephone:
             change['homephone'] = homephone
+    if 'user.chother' in __salt__ \
+            and other is not None:
```
Ok, I just did mimic the surrounding but you're right. All these should be one line comparison. I'm fixing it.

meaksh committed:
Do make comparisons in a single line

Commit:
Merge branch '2018.3' into 2018.3-remove-trailing-commas-on-linux-user-gecos-fields

rallytime commented:
@meaksh These test failures are related to this change: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/22849/#showFailuresLink

Can you take a look?

meaksh committed:
Add 'other' as valid kwargs for 'user.add' method

meaksh commented:
@rallytime - Oops! I missed those failures. They should be fixed now :+1:

Commit:
Merge branch '2018.3' into 2018.3-remove-trailing-commas-on-linux-user-gecos-fields

rallytime commented:
re-run py
